README.rst
async-rpc
=========

Non-blocking XML-RPC client for Python. Provides base classes
``async_rpc.BaseSerializer`` and ``async_rpc.BaseServerProxy``,
you can write your own implementation of RPC (JSON RPC, …).

Usage
-----

.. code-block:: python

    from async_rpc import XmlRpcServerProxy

    async def get_data():
        proxy = XmlRpcServerProxy('http://example.com/RPC2:8000')
        response = await proxy.methodName(123, 456)
        return response

API documentation
-----------------

*class* async_rpc.\ **BaseRpcServerProxy**\ (*uri, timeout=None,
max_clients=None, user_agent=None, use_dns_cache=None, ttl_dns_cache=None,
http_version=None, keepalive_timeout=None, \*\*kwargs*)

Base class for server proxies. It is responsible for HTTP transport. For
concrete RPC implementation uses descendant of the *BaseSerializer*.

- **uri** (*str*) – URL of the remote server
- **timeout** (*float*) – timeout in seconds for network operations,
  includes DNS requests, default is 1.0 seconds.
- **max_clients** (*int*) – size of the connections pool,
  default is 16 connections.
- **user_agent** (*str*) – user agent which is sent to server,
  default is ``"Python async-rpc"``.
- **use_dns_cache** (*bool*) – use internal cache for DNS lookups,
  default is ``True``.
- **ttl_dns_cache** (*float*) – store DNS lookups into internal cache
  for a *ttl_dns_cache* seconds, default is 10.0 seconds.
- **http_version** (*str*) – version of the HTTP protocol, can be
  ``"1.0"`` or ``"1.1"``, default is ``"1.0"``.
- **keepalive_timeout** (*float*) – close connection after *keepalive_timeout*
  seconds, if ``None``, keep-alive is disabled, default is ``None``.
- **kwargs**  (*dict*) – additional keyword arguments, which will be passed
  into serializer constructor.

*attribute* async_rpc.BaseRpcServerProxy.\ **serializer_cls**
(*BaseSerializer*) – serializer class

*coroutine* async_rpc.BaseRpcServerProxy.\ **call**\ (*name, \*\*params*)

- **name** (*str*) – mame of the remote function.
- **params**  (*dict*) – parameters, which will be passed to remote function.

Instance of the **BaseServerProxy** is callable, so you can call remote
function directly on server proxy instance. These calls are equivalent:

.. code-block:: python

    res = await proxy.getData('12345678901234567890')
    res = await proxy.call('getData', '12345678901234567890')


*class* async_rpc.\ **BaseSerializer**\ (*\*\*kwargs*)

Ancestor for concrete implementation of RPC. Contains four abstract methods,
which must be overriden in inherited classes.

- *kwargs* (*dict*) – additional keyword arguments, which are passed from
  server proxy.

*method* async_rpc.BaseSerializer.\ **initialize**\ (*\*\*kwargs*)

Initialize instance. It is called from constructor when instance is
created.

- **kwargs** (*dict*) – additional keyword arguments, which are passed from
  server proxy.

*abstractmethod* async_rpc.BaseSerializer.\ **prepare_request_headers**\ (
*params*)

Return ``dict`` containig HTTP headers. Method is called before RPC call
is sent. You can add additional HTTP header, which are be sent in request
to remote server.

- **params** (*tuple*) – params for RPC call.

*abstractmethod* async_rpc.BaseSerializer.\ **dumps**\ (*response_headers*)

Return data, which will be sent as POST body in request to remote
server. Method is called before RPC call is sent.

- **params** (*tuple*) – params for RPC call.
- **menthodname** (*str*) – name of the RPC method.

*abstractmethod* async_rpc.BaseSerializer.\ **process_response_headers**\ (
*response_headers*)

Check headers from response. Method is called after response is
received from server.

- **response_headers** (*Mapping*) – mapping containing response
  HTTP headers.

*abstractmethod* async_rpc.BaseSerializer.\ **loads**\ (*data,
response_headers*)

Return response from remote server as Python objects. Method is
called after response is received from server.

- **data** (*bytes*) – response body.
- **response_headers** (*Mapping*) – mapping containing response
  HTTP headers.

License
-------

3-clause BSD


CHANGELOG.md
# async-rpc

## [1.0.0] - 2018-05-19 16:23 - Jan Seifert <jan.seifert@fotkyzcest.net>
- First version of the *async-rpc*


tests/test_xmlrpc.py

import asyncio
import socket
import threading
import xmlrpc.server

import pytest

from async_rpc import XmlRpcServerProxy


@pytest.fixture(scope='module')
def xml_rpc_server():
    sock = socket.socket()
    sock.bind(('', 0))
    try:
        port = sock.getsockname()[1]
    finally:
        sock.close()

    def server_thread():
        server.serve_forever(poll_interval=0.1)

    server = xmlrpc.server.SimpleXMLRPCServer(
            ('localhost', port),
            requestHandler=xmlrpc.server.SimpleXMLRPCRequestHandler)
    try:
        server.register_introspection_functions()
        server.register_function(lambda a, b: a+b, 'sum')

        t = threading.Thread(target=server_thread, daemon=True)
        t.start()

        yield ('localhost', port)
    finally:
        server.shutdown()


@pytest.mark.asyncio
async def test_xml_rpc_server_proxy(xml_rpc_server):
    proxy = XmlRpcServerProxy('http://{}:{}/'.format(*xml_rpc_server))
    assert await proxy.sum(3, 7) == 10


@pytest.mark.asyncio
async def test_xml_rpc_server_proxy_fail_when_timeout(xml_rpc_server):
    proxy = XmlRpcServerProxy(
        'http://{}:{}/'.format(*xml_rpc_server), timeout=0.0)
    with pytest.raises(asyncio.TimeoutError):
        await proxy.sum(3, 7)


tests/test_serverproxy.py

from unittest import mock

import aiohttp
import pytest

from async_rpc import BaseSerializer, BaseServerProxy


class Serializer(BaseSerializer):

    def initialize(self, dummy_param=None):
        self.dummy_param = dummy_param

    def prepare_request_headers(self, unused_params):
        return {
            'Content-Type': 'application/x.foo',
            'Accept': 'application/x.foo',
        }

    def dumps(self, params, methodname):
        return {'method': methodname, 'args': params}

    def process_response_headers(self, response_headers):
        if response_headers['Content-Type'] != 'application/x.foo':
            raise ValueError('Invalid content type')

    def loads(self, data, response_headers):
        return (data['response'], response_headers['Content-Type'])


class ServerProxy(BaseServerProxy):

    serializer_cls = Serializer


@pytest.mark.parametrize(
    'proxy_kwargs, expected', [
        (
            {},
            {'timeout': 1.0, 'max_clients': 16, 'use_dns_cache': True,
             'ttl_dns_cache': 10.0, 'user_agent': 'Python async-rpc'}
        ),
        (
            {'timeout': 0.02, 'max_clients': 32,
             'user_agent': 'AasyncRPC',
             'use_dns_cache': False, 'ttl_dns_cache': 60.0},
            {'timeout': 0.02, 'max_clients': 32, 'use_dns_cache': False,
             'ttl_dns_cache': 60.0, 'user_agent': 'AasyncRPC'}
        ),
    ])
def test_base_server_proxy_constructor(proxy_kwargs, expected):
    proxy = ServerProxy('https://rpc.example.com/RPC2', **proxy_kwargs)
    assert proxy.uri == 'https://rpc.example.com/RPC2'
    assert proxy.host == 'rpc.example.com'
    assert proxy.timeout == expected['timeout']
    assert proxy.max_clients == expected['max_clients']
    assert proxy.user_agent == expected['user_agent']
    assert proxy.use_dns_cache is expected['use_dns_cache']
    assert proxy.ttl_dns_cache == expected['ttl_dns_cache']
    assert proxy.keepalive_timeout is None
    assert proxy.http_version == aiohttp.HttpVersion10
    assert isinstance(proxy.serializer, Serializer)


@pytest.mark.parametrize('http_version, expected', [
    (None, aiohttp.HttpVersion10), ('1.0', aiohttp.HttpVersion10),
    ('1.1', aiohttp.HttpVersion11)])
def test_base_server_proxy_constructor_http_version(http_version, expected):
    proxy = ServerProxy('https://example.com/RPC2', http_version=http_version)
    assert proxy.http_version == expected


def test_base_server_proxy_constructor_fail_when_invalid_http_version():
    with pytest.raises(ValueError):
        ServerProxy('https://example.com/RPC2', http_version='2.0')


def test_base_server_proxy_session():
    proxy = ServerProxy('https://rpc.example.com/RPC2')
    assert 'session' not in vars(proxy)
    with mock.patch('aiohttp.ClientSession') as m_client_session:
        with mock.patch('aiohttp.TCPConnector') as m_tcp_connector:
            with mock.patch('aiohttp.ClientTimeout') as m_client_timeout:
                session = proxy.session
    assert 'session' in vars(proxy)
    assert session == m_client_session.return_value
    m_client_timeout.assert_called_once_with()
    m_tcp_connector.assert_called_once_with(
        keepalive_timeout=None,
        force_close=True,
        limit=16,
        use_dns_cache=True,
        ttl_dns_cache=10)
    m_client_session.assert_called_once_with(
        connector=m_tcp_connector.return_value,
        version=aiohttp.HttpVersion10,
        timeout=m_client_timeout.return_value,
        raise_for_status=True,
        skip_auto_headers=['Accept-Encoding'])


@pytest.mark.asyncio
async def test_base_server_callable():
    async def call_m(name, *args):
        return (name, *args)
    proxy = ServerProxy('https://rpc.example.com/RPC2')
    with mock.patch.object(proxy, 'call', call_m):
        res = await proxy.sum(1, 2)
    assert res == ('sum', 1, 2)


@pytest.mark.asyncio
async def test_base_server_call_method():
    session_post_m_calls = []

    async def session_post(*args, **kwargs):
        async def response_read():
            return {
                'response': 3,
            }
        session_post_m_calls.append((args, kwargs))
        return mock.Mock(
            read=response_read,
            headers={'Content-Type': 'application/x.foo'},
        )
    session_post_m = mock.Mock(post=session_post)

    proxy = ServerProxy('https://rpc.example.com/RPC2')
    with mock.patch.dict(proxy.__dict__, {'session': session_post_m}):
        res = await proxy.call('sum', 1, 2)
    assert res == (3, 'application/x.foo')
    assert len(session_post_m_calls) == 1
    assert session_post_m_calls[0] == (
        ('https://rpc.example.com/RPC2',),
        {
            'data': {'method': 'sum', 'args': (1, 2)},
            'headers': {
                'User-Agent': 'Python async-rpc',
                'Host': 'rpc.example.com',
                'Content-Type': 'application/x.foo',
                'Accept': 'application/x.foo',
            },
        },
    )


async_rpc/xmlrpc.py

import xmlrpc.client

from .serverproxy import BaseSerializer, BaseServerProxy

__all__ = ['XmlRpcServerProxy']


class XmlRpcSerializer(BaseSerializer):

    def initialize(
            self, encoding=None, allow_none=False, use_builtin_types=False):
        self.encoding = 'utf8' if encoding is None else encoding
        self.allow_none = allow_none
        self.use_builtin_types = use_builtin_types

    def prepare_request_headers(self, unused_params):
        return {
            'Content-Type': 'text/xml',
            'Accept': 'text/xml',
        }

    def dumps(self, params, methodname):
        return xmlrpc.client.dumps(
            params, methodname=methodname, encoding=self.encoding,
            allow_none=self.allow_none)

    def process_response_headers(self, response_headers):
        ct = response_headers['Content-Type']
        if ct != 'text/xml':
            raise ValueError("Invalid response content type '{}'".format(ct))

    def loads(self, data, unused_response_headers):
        return xmlrpc.client.loads(
            data, use_builtin_types=self.use_builtin_types)[0][0]


class XmlRpcServerProxy(BaseServerProxy):

    serializer_cls = XmlRpcSerializer


async_rpc/__init__.py

import importlib
import pkgutil
import sys

from .serverproxy import BaseSerializer, BaseServerProxy
from .version import __version__ as _version
from .xmlrpc import XmlRpcServerProxy

__all__ = ['BaseSerializer', 'BaseServerProxy', 'XmlRpcServerProxy']

__version__ = _version

for extension_pkg_info in pkgutil.iter_modules():
    extension_name = extension_pkg_info[1]
    if not extension_name.startswith('async_rpc_'):
        del extension_name
        del extension_pkg_info
        continue
    extension_pkg = importlib.import_module(extension_name)
    for obj_name in getattr(extension_pkg, '__all__', []):
        async_rpc_pkg = sys.modules[__package__]
        setattr(async_rpc_pkg, obj_name, getattr(extension_pkg, obj_name))
        async_rpc_pkg.__all__.append(obj_name)
        del async_rpc_pkg
        del obj_name
    del extension_pkg
    del extension_name
    del extension_pkg_info


async_rpc/serverproxy.py

import abc
import functools
import urllib.parse

import aiohttp
import async_timeout

__all__ = ['BaseSerializer', 'BaseServerProxy']


class BaseSerializer(abc.ABC):

    def __init__(self, **kwargs):
        self.initialize(**kwargs)

    def initialize(self, **kwargs):
        pass

    @abc.abstractmethod
    def prepare_request_headers(self, params):
        pass

    @abc.abstractmethod
    def dumps(self, params, methodname):
        pass

    @abc.abstractmethod
    def process_response_headers(self, response_headers):
        pass

    @abc.abstractmethod
    def loads(self, data, response_headers):
        pass


class BaseServerProxy(object):

    serializer_cls = BaseSerializer

    _http_versions = {
        None: aiohttp.HttpVersion10,
        '1.0': aiohttp.HttpVersion10,
        '1.1': aiohttp.HttpVersion11,
    }

    def __init__(
            self, uri, timeout=None, max_clients=None, user_agent=None,
            use_dns_cache=None, ttl_dns_cache=None, http_version=None,
            keepalive_timeout=None, **kwargs):
        self.uri = uri
        self.host = urllib.parse.urlparse(uri).netloc
        self.timeout = 1.0 if timeout is None else timeout
        self.max_clients = 16 if max_clients is None else max_clients
        self.user_agent = (
            'Python async-rpc' if user_agent is None else user_agent)
        self.use_dns_cache = True if use_dns_cache is None else use_dns_cache
        self.ttl_dns_cache = 10.0 if ttl_dns_cache is None else ttl_dns_cache
        self.keepalive_timeout = keepalive_timeout
        try:
            self.http_version = self._http_versions[http_version]
        except KeyError:
            raise ValueError(
                'Unsupported HTTP version {}'.format(http_version))
        self.serializer = self.serializer_cls(**kwargs)

    @property
    def session(self):
        if 'session' not in self.__dict__:
            session = aiohttp.ClientSession(
                connector=aiohttp.TCPConnector(
                    keepalive_timeout=self.keepalive_timeout,
                    force_close=False if self.keepalive_timeout else True,
                    limit=self.max_clients,
                    use_dns_cache=self.use_dns_cache,
                    ttl_dns_cache=self.ttl_dns_cache,
                ),
                version=self.http_version,
                timeout=aiohttp.ClientTimeout(),
                raise_for_status=True,
                skip_auto_headers=['Accept-Encoding'],
            )
            self.__dict__['session'] = session
        return self.__dict__['session']

    def __getattr__(self, name):
        return functools.partial(self.call, name)

    async def call(self, name, *params):
        headers = {
            'User-Agent': self.user_agent,
            'Host': self.host,
        }
        request_headers = self.serializer.prepare_request_headers(params)
        if request_headers:
            headers.update(request_headers)
        data = self.serializer.dumps(params, name)
        async with async_timeout.timeout(self.timeout):
            response = await self.session.post(
                self.uri, data=data, headers=headers)
            data = await response.read()
        self.serializer.process_response_headers(response.headers)
        return self.serializer.loads(data, response.headers)
